### ISSUE TITLE
cmake building does not seem to work

### ISSUE THREAD

**Issue by 7er** [libjpeg-turbo/libjpeg-turbo#21]
I cloned the repo and did
> cmake .
> make

This results in:

<source-directory>/jchuff.c:380:2: error: #error Cannot determine word size
 #error Cannot determine word size
  ^

**Comment by dcommander** [comment]
Read BUILDING.txt.  CMake is used only for Windows builds.

**Comment by nathanaeljones** [comment]
Would cross-platform CMake changes be rejected? Before this repository came online, we were doing triple-platform (mac, linux, and windows) CI builds using Travis and AppVeyor - **with CMake** -> https://github.com/imazen/libjpeg-turbo

Changes between r1343 and  3395eddc936d99ea75b3802adde0ffe7cc4e8c23 have broken the build, although lack of shared history makes this very hard to debug.  

Our fork of FreeImage builds a lot of different projects, and we have found CMake to be essential in getting a consistent cross-platform build.

**Comment by dcommander** [comment]
Ah, so you weren't actually building from a "stock" version of libjpeg-turbo.  That makes more sense.  If you were building from a stock version, then CMake should have aborted with:

    Platform not supported by this build system.  Use autotools instead.

Unfortunately, at the moment I'm not receptive to adding more platforms to the CMake build system, because maintaining two build systems for the same platform is a royal PITA.  libjpeg-turbo uses a semi-continuous integration system, so build system errors are automatically tested, but if there are two build systems for the same platform, then one of those systems would not get tested.  If I had my choice, I'd move the entire project to CMake, but autotools is too heavily entrenched on Un*x, and libjpeg-turbo is too much of a core infrastructure library to do anything too radical in that department.  So it has always made the most sense to keep autotools on Un*x and use CMake on Windows only.

I think we can address the specific problem you're having, though.  The issue is that autotools defines SIZEOF_SIZE_T in jconfig.h, and the Huffman codec relies on this macro to determine whether it's being built with a 32-bit or a 64-bit compiler.  You would simply need to modify your CMakeLists.txt file to define that macro.  e.g.:

    add_definitions(-DSIZEOF_SIZE_T=${CMAKE_SIZEOF_VOID_P})

**Comment by nathanaeljones** [comment]
The only changes made were to cmakelists, actually - stock except for fixes.

> but if there are two build systems for the same platform, then one of those systems would not get tested. 

CI systems allow pivots; we test more than one build system for many of our projects. There are many forks trying to maintain xplat CMake for libjpeg-turbo - better if it's automatically tested, eh? 

> semi-continuous integration system

Thoughts on using Travis and AppVeyor so PRs are automatically tested?

**Comment by dcommander** [comment]
If an organization were to step forward and provide a general fund for this project, then yes.  Otherwise no.  Every minute I spend on it is costing me money, of which I have very little because I've already done many thousands of dollars of pro bono work for libjpeg-turbo this year alone.  I just can't add more features for free.

**Comment by nathanaeljones** [comment]
I've actually lobbied on your behalf with many corporations asking them to fund image codecs and improved real-time image processing, since everyone is in the same boat. So far no luck. Lots of short-sightedness going around. I'm tempted to do a Kickstarter.

The corporations that said no were: 
* Microsoft - this was a formal and endorsed proposal that went pretty far up the chain. 
* CloudFlare - they didn't rule it out but didn't want to invest anything at this time. I hope they change their minds and help improve the state of image processing on the web.
*  Bing - separately from MSFT.
* ... and several others. 

I've been trying to help the various projects that I benefit from (zlib, libpng, libjpeg-turbo, libtiff, libraw, freeimage, libgd, etc) reduce maintenance costs by helping them move to tools like GitHub, Travis, AppVeyor, Coverity, etc.  All of our free cash at Imazen the last 3 years has gone towards those efforts, but we're dry at this point, thanks to MSFT undermining our platform (.NET is our business model). 

The offer is to get Travis and AppVeyor set up for you to reduce maintenance costs. I think this would be an easy win, given how it's helped us on other projects.

**Comment by dcommander** [comment]
I greatly appreciate the lobbying efforts.  Let's hope someone steps forward at some point.

BTW, the actual commit that introduced the SIZEOF_SIZE_T check was 63ba814364b465e03a20d0240cf.  Due to adding a bunch of fake merge info when we migrated to GitHub (at the request of Mozilla), for some reason 'git blame' isn't properly working with that particular line.

I appreciate that those platforms will probably help me reduce maintenance costs in the long term, but in the near term I can't spare the cycles to migrate to them and learn them.  I already spent a great deal of time in the past few months migrating to and getting familiar with GitHub (more thousands of dollars in unpaid labor.)

For the moment at least, you're going to have to work around this in your own build system.

**Comment by dcommander** [comment]
Cross-platform CMake build system is now available in dev.  Please test.

**Comment by TamaMcGlinn** [comment]
I'm getting the same error when trying to build [minetest-wasm](https://github.com/paradust7/minetest-wasm), which uses this library. The full error is:

```
/home/tama/code/computercraft/minetest-wasm/sources/libjpeg/jchuff.c:92:46: error: invalid token at start of a preprocessor expression
#if (defined(SIZEOF_SIZE_T) && SIZEOF_SIZE_T == 8) || defined(_WIN64) || \
                                             ^
/home/tama/code/computercraft/minetest-wasm/sources/libjpeg/jchuff.c:95:48: error: invalid token at start of a preprocessor expression
#elif (defined(SIZEOF_SIZE_T) && SIZEOF_SIZE_T == 4) || defined(_WIN32)
                                               ^
/home/tama/code/computercraft/minetest-wasm/sources/libjpeg/jchuff.c:98:2: error: Cannot determine word size
#error Cannot determine word size
 ^
```

minetest-wasm uses commit [2ee7264d40910f2529690de327988ce0c2276812](https://github.com/libjpeg-turbo/libjpeg-turbo/commit/2ee7264d40910f2529690de327988ce0c2276812), which is a descendant of your fix commit [6abd39160c5a3762e9ebe024e75407665093e715](https://github.com/libjpeg-turbo/libjpeg-turbo/commit/6abd39160c5a3762e9ebe024e75407665093e715). Unsurprisingly, when switching to that fix commit directly the error is still there:

```
/home/tama/code/computercraft/minetest-wasm/sources/libjpeg/jchuff.c:390:18: error: invalid token at start of a preprocessor expression
#if SIZEOF_SIZE_T==8 || define
```

I also tried adding a definition for SIZEOF_SIZE_T to CMakeLists.txt (line 8) as follows, but got the same result:

```
set(SIZEOF_SIZE_T 8)
```

**Comment by dcommander** [comment]
@TamaMcGlinn This issue is 7 years old, so I doubt you are seeing literally the same issue. This issue was originally posted by someone who was trying to use CMake to build libjpeg-turbo for Un*x, which wasn't supported at the time (but now is.) Your issue appears to be related to WASM. I am receptive to patches that better enable WASM support, but those patches will need to come from the community. I would suggest discussing WASM issues under #250.
